runtime/lib/mirrors_impl.dart

```diff
@@ -1664,15 +1664,35 @@ class _Mirrors {
     return classMirror;
   }
 
-  static TypeMirror reflectType(Type key) {
-    var typeMirror = _instantiationCache[key];
-    if (typeMirror == null) {
-      typeMirror = makeLocalTypeMirror(key);
-      _instantiationCache[key] = typeMirror;
-      if (typeMirror is ClassMirror && !typeMirror._isGeneric) {
-        _declarationCache[key] = typeMirror;
+  static TypeMirror reflectType(Type key, [Iterable<Type> typeArguments]) {
+    if (typeArguments == null || typeArguments.isEmpty) {
+      var typeMirror = _instantiationCache[key];
+      if (typeMirror == null) {
+        typeMirror = makeLocalTypeMirror(key, arguments);
+        _instantiationCache[key] = typeMirror;
+        if (typeMirror is ClassMirror && !typeMirror._isGeneric) {
+          _declarationCache[key] = typeMirror;
+        }
       }
+      return typeMirror;
+    } else {
+      // We can't cache type provided in `key` since actual type will depend
+      // on resolving provided `key` and type arguments.
+      typeArguments ??= [];
+      List arguments = typeArguments.toList(growable: false);
+      var resolvedTypeMirror = makeLocalTypeMirror(key, arguments);
+      var actualType = resolvedTypeMirror.reflectedType;
+
+      // We check if actual type has been cached already and return cached copy.
+      var cachedTypeMirror = _instantiationCache[actualType];
+      if (cachedTypeMirror == null) {
+        _instantiationCache[actualType] = resolvedTypeMirror;
+        if (resolvedTypeMirror is ClassMirror && !resolvedTypeMirror._isGeneric) {
```

Comment:
Can this ever be true?
In theory, the `resolvedTypeMirror` must be a generic class, since could provide generic arguments.